examples/nimble_gnrc/ble.c

```diff
@@ -0,0 +1,280 @@
+        .itvl_max = APP_ADV_ITVL,
+    };
+
+    /* make sure no advertising is in progress */
+    if (nimble_netif_conn_is_adv()) {
+        puts("err: advertising already in progress");
+        return;
+    }
+
+    /* build advertising data */
+    res = bluetil_ad_init_with_flags(&ad, buf, BLE_HS_ADV_MAX_SZ,
+                                     BLUETIL_AD_FLAGS_DEFAULT);
+    assert(res == BLUETIL_AD_OK);
+    uint16_t ipss = BLE_SVC_IPSS;
+    res = bluetil_ad_add(&ad, BLE_GAP_AD_UUID16_INCOMP, &ipss, sizeof(ipss));
+    assert(res == BLUETIL_AD_OK);
+    if (name == NULL) {
+        name = APP_ADV_NAME_DEFAULT;
+    }
+    res = bluetil_ad_add(&ad, BLE_GAP_AD_NAME, name, strlen(name));
+    if (res != BLUETIL_AD_OK) {
+        puts("err: the given name is too long");
+        return;
+    }
+
+    /* start listening for incoming connections */
+    res = nimble_netif_accept(ad.buf, ad.pos, &_adv_params);
+    if (res != NIMBLE_NETIF_OK) {
+        printf("err: unable to start advertising (%i)\n", res);
+    }
+    else {
+        printf("success: advertising this node as '%s'\n", name);
+    }
+}
+
+static void _cmd_adv_stop(void)
+{
+    int res = nimble_netif_accept_stop();
+    if (res == NIMBLE_NETIF_OK) {
+        puts("canceled advertising");
+    }
+    else if (res == NIMBLE_NETIF_NOTADV) {
+        puts("no advertising in progress");
+    }
+}
+
+static void _cmd_scan(unsigned duration)
+{
+    if (duration == 0) {
+        return;
+    }
+    printf("scanning (for %ums) ... ", (duration / 1000));
+    nimble_scanlist_clear();
+    nimble_scanner_start();
+    xtimer_usleep(duration);
+    nimble_scanner_stop();
+    puts("done");
+    nimble_scanlist_print();
+}
+
+static void _cmd_connect(unsigned pos)
+{
+    nimble_scanlist_entry_t *sle = nimble_scanlist_get_by_pos(pos);
+    if (sle == NULL) {
+        puts("err: unable to find given entry in scanlist");
+        return;
+    }
+
+    // TODO: for now we use default parameters
+    int res = nimble_netif_connect(&sle->addr, NULL, APP_CONN_TIMEOUT);
+    if (res < 0) {
+        printf("err: unable to trigger connection sequence (%i)\n", res);
+        return;
+    }
+
+    printf("initiated connection procedure with ");
+    bluetil_addr_print(sle->addr.val);
+    puts("");
+}
+
+static void _cmd_close(int handle)
+{
+    int res = nimble_netif_close(handle);
+    if (res != NIMBLE_NETIF_OK) {
+        puts("err: unable to close connection with given handle");
+    }
+    else {
+        puts("success: connection tear down initiated");
+    }
+}
+
+static int _ishelp(char *argv)
+{
+    return memcmp(argv, "help", 4) == 0;
+}
+
+void app_ble_init(void)
+{
+    /* setup the scanning environment */
+    nimble_scanlist_init();
+    nimble_scanner_init(NULL, nimble_scanlist_update);
+
+    /* register event callback with the netif wrapper */
+    nimble_netif_eventcb(_on_ble_evt);
+}
+
+int app_ble_cmd(int argc, char **argv)
+{
+    if ((argc == 1) || _ishelp(argv[1])) {
```

Comment:
A bit crowded and hard to read. Might be a good application for getopt().

Reply:
true, but I would take this as optimization that might be done in the future. This is IMO not the most relevant code path anyway.